From fbaa23a4beed12d831b50da699a5ad975a56848e Mon Sep 17 00:00:00 2001
From: David Milburn <dmilburn@redhat.com>
Date: Mon, 18 Aug 2014 19:58:43 -0400
Subject: [block] nvme: Fix the buffer size passed in GetLogPage(CDW10.NUMD)

Message-id: <1408391935-24886-11-git-send-email-dmilburn@redhat.com>
Patchwork-id: 87891
O-Subject: [RHEL7.1 PATCH BZ 1111259 10/22] NVMe: Fix the buffer size passed in GetLogPage(CDW10.NUMD)
Bugzilla: 1111259
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Indraneel Mukherjee <indraneel.m@samsung.com>

In GetLogPage the buffer size passed to device is a 0's based value.

Signed-off-by: Indraneel M <indraneel.m@samsung.com>
Reported-by: Shiro Itou <shiro.itou@outlook.com>
Reviewed-by: Vishal Verma <vishal.l.verma@linux.intel.com>
Signed-off-by: Matthew Wilcox <matthew.r.wilcox@intel.com>
(cherry picked from commit 4131f2fcdc5853d995044d4ed995a25ee4eb5ab2)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/block/nvme-scsi.c b/drivers/block/nvme-scsi.c
index da3b252..7ec61c4 100644
--- a/drivers/block/nvme-scsi.c
+++ b/drivers/block/nvme-scsi.c
@@ -1014,8 +1014,8 @@ static int nvme_trans_log_info_exceptions(struct nvme_ns *ns,
  c.common.opcode = nvme_admin_get_log_page;
  c.common.nsid = cpu_to_le32(0xFFFFFFFF);
  c.common.prp1 = cpu_to_le64(dma_addr);
- c.common.cdw10[0] = cpu_to_le32(((sizeof(struct nvme_smart_log) /
-   BYTES_TO_DWORDS) << 16) | NVME_GET_SMART_LOG_PAGE);
+ c.common.cdw10[0] = cpu_to_le32((((sizeof(struct nvme_smart_log) /
+   BYTES_TO_DWORDS) - 1) << 16) | NVME_GET_SMART_LOG_PAGE);
  res = nvme_submit_admin_cmd(dev, &c, NULL);
  if (res != NVME_SC_SUCCESS) {
   temp_c = LOG_TEMP_UNKNOWN;
@@ -1082,8 +1082,8 @@ static int nvme_trans_log_temperature(struct nvme_ns *ns, struct sg_io_hdr *hdr,
  c.common.opcode = nvme_admin_get_log_page;
  c.common.nsid = cpu_to_le32(0xFFFFFFFF);
  c.common.prp1 = cpu_to_le64(dma_addr);
- c.common.cdw10[0] = cpu_to_le32(((sizeof(struct nvme_smart_log) /
-   BYTES_TO_DWORDS) << 16) | NVME_GET_SMART_LOG_PAGE);
+ c.common.cdw10[0] = cpu_to_le32((((sizeof(struct nvme_smart_log) /
+   BYTES_TO_DWORDS) - 1) << 16) | NVME_GET_SMART_LOG_PAGE);
  res = nvme_submit_admin_cmd(dev, &c, NULL);
  if (res != NVME_SC_SUCCESS) {
   temp_c_cur = LOG_TEMP_UNKNOWN;
-- 
1.7.1